specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2018-02-01-preview/containerregistry_build.json

```diff
@@ -0,0 +1,2089 @@
+          }
+        },
+        "name": {
+          "description": "The name of the argument.",
+          "type": "string"
+        },
+        "value": {
+          "description": "The value of the argument.",
+          "type": "string"
+        },
+        "isSecret": {
+          "description": "Flag to indicate whether the argument represents a secret and want to be removed from build logs.",
+          "default": false,
+          "type": "boolean"
+        }
+      }
+    },
+    "BaseImageDependencyList": {
+      "type": "object",
+      "properties": {
+        "value": {
+          "description": "The collection value.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/BaseImageDependency"
+          }
+        },
+        "nextLink": {
+          "description": "The URI that can be used to request the next set of paged results.",
+          "type": "string"
+        }
+      }
+    },
+    "BaseImageDependency": {
+      "description": "Properties that describe a base image dependency.",
+      "required": [
+        "repositoryName",
+        "digest"
+      ],
+      "type": "object",
+      "properties": {
+        "type": {
+          "description": "The type of the base image dependency.",
+          "enum": [
+            "BuildTime",
+            "RunTime"
+          ],
+          "type": "string",
+          "x-ms-enum": {
+            "name": "BaseImageDependencyType",
+            "modelAsString": true
+          }
+        },
+        "isAutoTriggerEnabled": {
+          "description": "The value of this flag indicates if build has to be auto triggered when this image is updated.",
+          "default": false,
+          "type": "boolean"
+        },
+        "repositoryName": {
+          "description": "Name of the repository.",
+          "type": "string"
+        },
+        "tag": {
+          "description": "The tag name.",
+          "type": "string"
+        },
+        "digest": {
+          "description": "The image digest. Example: SHA256 based digest.",
+          "type": "string"
+        }
+      }
+    },
+    "QueueBuildRequest": {
+      "description": "The queue build request parameters.",
+      "type": "object",
+      "properties": {
+        "type": {
+          "description": "The type of the build request.",
+          "type": "string",
+          "readOnly": true
+        }
+      },
+      "discriminator": "type"
+    },
+    "SourceUploadDefinition": {
+      "description": "The properties of a response to source upload request.",
+      "type": "object",
+      "properties": {
+        "uploadUrl": {
+          "description": "The URL where the client can upload the source.",
+          "type": "string"
+        },
+        "relativePath": {
+          "description": "The relative path to the source. This is used to submit the subsequent queue build request.",
+          "type": "string"
+        }
+      }
+    },
+    "Resource": {
```

Comment:
It would be better to define a `Resource` with 3 models, like this...
```json
   "Resource": {
      "description": "The core properties of ARM resources",
      "properties": {
        "id": {
          "readOnly": true,
          "type": "string",
          "description": "Fully qualified resource Id for the resource"
        },
        "name": {
          "readOnly": true,
          "type": "string",
          "description": "The name of the resource"
        },
        "type": {
          "readOnly": true,
          "type": "string",
          "description": "The type of the resource."
        }
      },
      "x-ms-azure-resource": true
    },
    "TrackedResource": {
      "description": "The resource model definition for a ARM tracked top level resource",
      "properties": {
        "tags": {
          "type": "object",
          "additionalProperties": {
            "type": "string"
          },
          "x-ms-mutability": [
            "read",
            "create",
            "update"
          ],
          "description": "Resource tags."
        },
        "location": {
          "type": "string",
          "x-ms-mutability": [
            "read",
            "create"
          ],
          "description": "The Azure Region where the resource lives"
        }
      },
      "allOf": [
        {
          "$ref": "#/definitions/Resource"
        }
      ]
    },
    "ProxyResource": {
      "description": "The resource model definition for a ARM proxy resource. It will have everything other than required location and tags",
      "allOf": [
        {
          "$ref": "#/definitions/Resource"
        }
      ]
    },
```

And, when defining models that are actually azure resources (like `Build`, `BuildDefinition` and `BuildStep`), derive them from `TrackedResource` instead of just `Resource`

Reply:
Not all of them are tracked resource. BuildDefinition is the only tracked resource. Others are REST resources.

Reply:
I see, will take a look.

Reply:
I fixed it.

Reply:
Err, no definition for tracked resource :( ?

---
In reply to: [174971511](https://github.com/Azure/azure-rest-api-specs/pull/2378#discussion_r174971511) [](ancestors = 174971511)